### PR TITLE
RFR - When private_key is supplied as runner param, use that key for auth. 

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
## Examples

```
/m/s/s/st2 git:issue_1889_private_key ❯❯❯ st2 run core.remote cmd=whoami hosts=localhost                                ✭ ✱ ◼
.
id: 55dff0bd32ed356c736318b0
status: succeeded
result:
{
    "localhost": {
        "succeeded": true,
        "failed": false,
        "return_code": 0,
        "stderr": "",
        "stdout": "lakshmi
"
    }
}
/m/s/s/st2 git:issue_1889_private_key ❯❯❯ st2 run core.remote cmd=whoami hosts=localhost username=stanley               ✭ ✱ ◼
.
id: 55dff0c332ed356c736318b3
status: failed
result:
{
    "traceback": "  File "/mnt/src/storm/st2/st2actions/st2actions/container/base.py", line 99, in _do_run
    runner.pre_run()
  File "/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/paramiko_ssh_runner.py", line 93, in pre_run
    raise InvalidCredentialsException(msg)
",
    "error": "Either password or private_key data needs to be supplied for user: stanley"
}
/m/s/s/st2 git:issue_1889_private_key ❯❯❯ st2 run core.remote cmd=whoami hosts=localhost username=stanley private_key=/home/vagrant/.ssh/stanley_rsa
.
id: 55dff0d132ed356c736318b6
status: failed
result:
{
    "traceback": "  File "/mnt/src/storm/st2/st2actions/st2actions/container/base.py", line 99, in _do_run
    runner.pre_run()
  File "/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/paramiko_ssh_runner.py", line 130, in pre_run
    connect=True
  File "/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/parallel_ssh.py", line 55, in __init__
    connect_results = self.connect(raise_on_any_error=raise_on_any_error)
  File "/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/parallel_ssh.py", line 85, in connect
    raise NoHostsConnectedToException(msg)
",
    "error": "Unable to connect to any one of the hosts: [u'localhost'].

 connect_errors={
  "localhost": {
    "failed": true,
    "traceback": "  File \"/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/parallel_ssh.py\", line 239, in _connect\n    client.connect()\n  File \"/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/paramiko_ssh.py\", line 121, in connect\n    conninfo['pkey'] = self._get_pkey_object(key=self.key_material)\n  File \"/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/paramiko_ssh.py\", line 473, in _get_pkey_object\n    raise paramiko.ssh_exception.SSHException(msg)\n",
    "return_code": 255,
    "succeeded": false,
    "error": "Cannot connect to host. Invalid or unsupported key type"
  }
}"
}
/m/s/s/st2 git:issue_1889_private_key ❯❯❯ st2 run core.remote cmd=whoami hosts=localhost username=stanley private_key="`cat /home/vagrant/.ssh/stanley_rsa`"
.
id: 55dff0de32ed356c736318b9
status: succeeded
result:
{
    "localhost": {
        "succeeded": true,
        "failed": false,
        "return_code": 0,
        "stderr": "",
        "stdout": "stanley
"
    }
}
/m/s/s/st2 git:issue_1889_private_key ❯❯❯
```

## TODO
[X] - Tests
[X] - Docs

### CODE REVIEW

**Comment by manasdk** [comment]
So if a user's private key is present on StackStorm box any user with access to StackStorm can execute an action on a remote box as that user? 

Are we sure this is a good idea?

**Comment by lakshmi-kannan** [comment]
The public key still needs to be deployed to the boxes that the users are connecting to. Only then, authentication will succeed.

private_key will be in the box where the client is. I am hoping people don't share their private_keys. Otherwise, this will be a disaster anyways.

**Comment by lakshmi-kannan** [comment]
Fabric behavior is bizzare! Look at the executions with fabric with username & username + private_key provided. https://gist.github.com/lakshmi-kannan/d22e697ea032b3f7d4d0

The intended behavior is to use the private_key provided but I think there might be a bug there.
https://github.com/StackStorm/st2/blob/master/st2common/st2common/models/system/action.py#L450

**Comment by manasdk** [comment]
> private_key will be in the box where the client is.

`st2 run examples.remote-fib hosts=localhost start=1 end=10 username=stanley private_key=/home/vagrant/.ssh/stanley_rsa`

The key is read by the CLI? We actually have some special handling in the CLI for this file?

**Comment by manasdk** [comment]
> Fabric behavior is bizzare! Look at the executions with fabric with username & username + private_key provided. https://gist.github.com/lakshmi-kannan/d22e697ea032b3f7d4d0

https://github.com/StackStorm/st2/blob/master/st2actions/st2actions/runners/ssh/fabric_runner.py#L46

Not really it is exactly as coded.

**Comment by lakshmi-kannan** [comment]
>The key is read by the CLI? We actually have some special handling in the CLI for this file?

I misunderstood this part. I am confused what's the behavior we want. So yeah, let's talk offline about this and then I'll open a PR.

**Comment by manasdk** [comment]
> I suggest you refactor it breaking current troubleshooting.rst in blocks

Lets do that in subsequent PRs please.

**Comment by manasdk** [comment]
@lakshmi-kannan This looks like it is still WIP. I don't expect us to ship this for 0.13.1 any longer.

**Comment by lakshmi-kannan** [comment]
@manasdk Not for 0.13.1. I have to add some tests and get the docs changed.

**Comment by lakshmi-kannan** [comment]
Unfortunately, this PR also touched docs and also other things. Apologies for the giant PR.

**Comment by Kami** [comment]
A bit late to the game, but:

> The key is read by the CLI? We actually have some special handling in the CLI for the private_key property?

It's only considered a file path and read if you prefix argument name with "@" (e.g. `@private_key=/path/to/key/file`). We follow similar convention as curl and also do the same for HTTP runner.

I believe this also documented in the CLI 101 section.

File: docs/source/troubleshooting/ssh.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+would require you to setup private_key for the users on stackstorm action runner boxes and
+the public keys of the users in target boxes. This increases the surface area for risk and
+is highly discouraged.
+
+Said that, if you have st2client installed and want to run one off commands on remote
+boxes as a different user, we have a way.
+
+::
+
+    $st2 run core.remote cmd=whoami hosts=localhost username=test_user private_key="`cat /home/vagrant/.ssh/id_rsa`"
```
Comment (Kami) [inline docs/source/troubleshooting/ssh.rst]: Noticed this a bit late, should just use "@" notation - `@private_key=/home/vagrant/.ssh/id_rsa`

**Comment by jfryman** [comment]
Hey @lakshmi-kannan 

Somehow in the confusion, I missed this. This really isn't ideal. Read up on some rant.

https://stackstorm.slack.com/archives/stackstorm/p1445294185009809

Let's see how we can still provide what the user needs, but in a safer way

**Pull request merged**